src/Equinox.CosmosStore/CosmosStore.fs

```diff
@@ -694,16 +712,11 @@ module internal Query =
     [<RequireQualifiedAccess; NoComparison; NoEquality>]
     type ScanResult<'event> = { found: bool; minIndex: int64; next: int64; maybeTipPos: Position option; events: 'event[] }
 
-    let scanTip (tryDecode: #IEventData<EventBody> -> 'event voption, isOrigin: 'event -> bool) (pos: Position, i: int64, xs: #ITimelineEvent<EventBody>[]): ScanResult<'event> =
-        let items = ResizeArray()
-        let isOrigin' e =
-            match tryDecode e with
-            | ValueNone -> false
-            | ValueSome e ->
-                items.Insert(0, e) // WalkResult always renders events ordered correctly - here we're aiming to align with Enum.EventsAndUnfolds
-                isOrigin e
-        let f, e = xs |> Seq.tryFindBack isOrigin' |> Option.isSome, items.ToArray()
-        { found = f; maybeTipPos = Some pos; minIndex = i; next = pos.index + 1L; events = e }
+    let scanTip (tryDecode, isOrigin) (pos: Position, i: int64, xs: #ITimelineEvent<EventBody>[]): ScanResult<'event> =
+        let ok, e = xs |> Tip.tryFindOrigin (tryDecode, isOrigin) |> function
+            | ValueSome (_, events) -> true, events
+            | ValueNone -> false, Array.empty
```

Comment:
maybe `tryFindOrigin` should return a `bool * 'E[]`?

Reply:
Hm #434 makes it less obvious as we use the value in another case
also if it wasnt for the ValueSome vs Some, there'd be less debate